xla/pjrt/pjrt_stream_executor_client.cc

```diff
@@ -1255,14 +1259,43 @@ Status PjRtStreamExecutorDevice::TransferFromOutfeed(
       local_device->local_hardware_id().value(), literal);
 }
 
+void PjRtStreamExecutorClient::SetMemorySpaces(PjRtDevice* device) {
+  if (platform_name() == "cuda") {
+    device_space_ = std::make_unique<DeviceMemorySpace>(
+        static_cast<int>(Layout::kDefaultMemorySpace), device);
+    nccl_space_ = std::make_unique<NcclMemorySpace>(
+        static_cast<int>(Layout::kGenericFastMemorySpace), device);
+    pinned_space_ = std::make_unique<PinnedHostMemorySpace>(
+        static_cast<int>(Layout::kHostMemorySpace), device);
+    unpinned_space_ = std::make_unique<UnpinnedHostMemorySpace>(
+        static_cast<int>(Layout::kHostMemorySpace), device);
+
+    memory_spaces_.clear();
+    memory_spaces_.emplace_back(
+        static_cast<PjRtMemorySpace*>(device_space_.get()));
+    memory_spaces_.emplace_back(
+        static_cast<PjRtMemorySpace*>(nccl_space_.get()));
+    memory_spaces_.emplace_back(
+        static_cast<PjRtMemorySpace*>(pinned_space_.get()));
+    memory_spaces_.emplace_back(
+        static_cast<PjRtMemorySpace*>(unpinned_space_.get()));
+  }
+
+}
+
 absl::Span<PjRtMemorySpace* const> PjRtStreamExecutorDevice::memory_spaces()
     const {
-  return {};
+  return client()->memory_spaces();
 }
 
 StatusOr<PjRtMemorySpace*> PjRtStreamExecutorDevice::default_memory_space()
     const {
-  return Unimplemented("default_memory_space is not supported.");
+  if (client_->platform_name() != "cuda") {
+    return Unimplemented(
+        "default_memory_space is not supported for platform '%s'.",
+        platform_name());
+  }
+  return memory_spaces()[0];
```

Comment:
I would suggest using a lookup or something instead of using that magic index 0. 
The `default_memory_space` can be changed when the order of emplace_back shifts?